### PR TITLE
Non blocking queue

### PR DESCRIPTION
**Pull request by wtrocki**
## Motivation

Avoid blocking offline queue. 
General idea is to return an error to the user instantly
and then trigger another mutation that will be enqueued as an offline one.

1. The request comes to an offline queue.
2. Return observable with error instantly (mutate().catch() will be triggered)
3. Attach observable to error object.

```javascript
client.mutate.try()
.catch(error => {
    error.offlineObservable.then((data)=>{
        // Successfully mutated after error
    })
})
```

Screen Shot 2019-04-12 at 5 56 41 PM

## Tasks

- [x] Implement nonblocking behavior
- [ ] Unify storage

### CODE REVIEW

File: packages/sync/src/links/LinksBuilder.ts [hunk 1]
```diff
@@ -8,10 +8,10 @@ import { AuditLoggingLink } from "./AuditLoggingLink";
 import { DefaultMetricsBuilder, MetricsBuilder } from "@aerogear/core";
 import { LocalDirectiveFilterLink } from "./LocalDirectiveFilterLink";
 import { createUploadLink } from "apollo-upload-client";
-import { isMutation, isOnlineOnly, isSubscription, isMarkedOffline } from "../utils/helpers";
+import { isMutation, isOnlineOnly, isSubscription } from "../utils/helpers";
 import { defaultWebSocketLink } from "./WebsocketLink";
 import { OfflineLink } from "./OfflineLink";
-import { NetworkStatus } from "../offline";
+import { NetworkStatus, OfflineRestoreHandler } from "../offline";
 import { extensionsLink } from "./ExtensionsLink";
```
Comment (jhellar) [inline packages/sync/src/links/LinksBuilder.ts]: This import can be removed.

File: packages/sync/src/links/OfflineLink.ts [hunk 2]
```diff
@@ -37,11 +43,30 @@ export class OfflineLink extends ApolloLink {
   }
 
   public request(operation: Operation, forward: NextLink) {
-    if (!this.online || isMarkedOffline(operation)) {
+    // Reattempting operation that was marked as offline
+    if (OfflineRestoreHandler.isMarkedOffline(operation)) {
+      logger.log("Enqueueing offline mutation", operation.variables);
```
Comment (jhellar) [inline packages/sync/src/links/OfflineLink.ts]: `logger.log(...)` -> `logger(...)`

File: packages/sync/src/links/OfflineLink.ts [hunk 3]
```diff
@@ -37,11 +43,30 @@ export class OfflineLink extends ApolloLink {
...
       return this.queue.enqueue(operation, forward);
     }
-    // We are online and can skip this link;
-    return forward(operation);
+
+    if (this.online) {
+      logger.log("Online: Forwarding mutation", operation.variables);
```
Comment (jhellar) [inline packages/sync/src/links/OfflineLink.ts]: `logger.log(...)` -> `logger(...)`

File: packages/sync/src/links/OfflineLink.ts [hunk 4]
```diff
@@ -37,11 +43,30 @@ export class OfflineLink extends ApolloLink {
...
+      // We are online and can skip this link;
+      return forward(operation);
+    }
+
+    if (!this.offlineMutationHandler) {
+      logger.log("Error: Offline link setup method was not called");
```
Comment (jhellar) [inline packages/sync/src/links/OfflineLink.ts]: `logger.log(...)` -> `logger(...)`

File: packages/sync/src/links/OfflineLink.ts [hunk 5]
```diff
@@ -37,11 +43,30 @@ export class OfflineLink extends ApolloLink {
...
+      return forward(operation);
+    }
+
+    const operationEntry = new OperationQueueEntry(operation, forward);
+    this.offlineMutationHandler.mutateOfflineElement(operationEntry);
+    return new Observable(observer => {
+      logger.log("Returning error to client", operation.variables);
```
Comment (jhellar) [inline packages/sync/src/links/OfflineLink.ts]: `logger.log(...)` -> `logger(...)`

File: packages/sync/src/offline/OfflineRestoreHandler.ts [hunk 6]
```diff
@@ -47,27 +48,52 @@ export class OfflineRestoreHandler {
...
-        mutation: item.operation.query,
-        // Restore optimistic response from operation in order to see it
-        optimisticResponse,
-        // Pass client update functions
-        update: updateFunction,
-        // Pass extensions as part of the context
-        context: { extensions }
-      };
-      this.apolloClient.mutate(mutationOptions);
+      await this.mutateOfflineElement(item);
```
Comment (jhellar) [inline packages/sync/src/offline/OfflineRestoreHandler.ts]: `await` should probably not be there. When removed, integration tests pass.

Comment (wtrocki) [reply packages/sync/src/offline/OfflineRestoreHandler.ts]: Good catch!

**Comment by StephenCoady** [comment]
@wtrocki took a look at the code and it looks great to me. I played with it too and its working great. When you make this ready for review I will review it fully!

File: packages/sync/src/offline/OfflineRestoreHandler.ts (hunk 6 above)
Comment (wtrocki) [reply packages/sync/src/offline/OfflineRestoreHandler.ts]: Actually going to address that so we will have async here.

**Comment by wtrocki** [comment]
@StephenCoady I'm going to record video on this and produce some diagrams to explain this idea. This will be the new architecture so the change will require major bump.

**Comment by wtrocki** [comment]
Ready for testing but integration tests worry me now a lot.

**Comment by wtrocki** [comment]
Ready for verification.
I have played with that on IOS and it works like a dream.
Outstanding part will be to write documentation. I will add documentation to separate repository (and release dev version now)

**Pull request merged**